<a href="https://colab.research.google.com/github/ankitsd/avidya/blob/master/try2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!wget https://datahack-prod.s3.amazonaws.com/test_file/test_Jwt0MQH.zip

--2019-05-04 20:50:53--  https://datahack-prod.s3.amazonaws.com/test_file/test_Jwt0MQH.zip
Resolving datahack-prod.s3.amazonaws.com (datahack-prod.s3.amazonaws.com)... 52.219.62.76
Connecting to datahack-prod.s3.amazonaws.com (datahack-prod.s3.amazonaws.com)|52.219.62.76|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 10810428 (10M) [application/zip]
Saving to: ‘test_Jwt0MQH.zip.3’

test_Jwt0MQH.zip.3  100%[===================>]  10.31M  4.49MB/s    in 2.3s    

2019-05-04 20:50:56 (4.49 MB/s) - ‘test_Jwt0MQH.zip.3’ saved [10810428/10810428]



In [2]:
!wget https://datahack-prod.s3.amazonaws.com/train_file/train_5CLrC8b.zip

--2019-05-04 20:50:59--  https://datahack-prod.s3.amazonaws.com/train_file/train_5CLrC8b.zip
Resolving datahack-prod.s3.amazonaws.com (datahack-prod.s3.amazonaws.com)... 52.219.64.12
Connecting to datahack-prod.s3.amazonaws.com (datahack-prod.s3.amazonaws.com)|52.219.64.12|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 27621619 (26M) [application/zip]
Saving to: ‘train_5CLrC8b.zip.3’

train_5CLrC8b.zip.3 100%[===================>]  26.34M  6.62MB/s    in 4.0s    

2019-05-04 20:51:04 (6.62 MB/s) - ‘train_5CLrC8b.zip.3’ saved [27621619/27621619]



In [3]:
!unzip train_5CLrC8b.zip
!unzip test_Jwt0MQH.zip

Archive:  train_5CLrC8b.zip
replace Data_Dictionary.xlsx? [y]es, [n]o, [A]ll, [N]one, [r]ename: y
  inflating: Data_Dictionary.xlsx    
replace train.csv? [y]es, [n]o, [A]ll, [N]one, [r]ename: y
  inflating: train.csv               
Archive:  test_Jwt0MQH.zip
replace test.csv? [y]es, [n]o, [A]ll, [N]one, [r]ename: y
  inflating: test.csv                


In [0]:
import pandas as pd
from datetime import datetime
dateparse = lambda x: datetime.strptime(x, '%d-%m-%Y')

In [5]:
df_train = pd.read_csv("train.csv",parse_dates=True,date_parser=dateparse)
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 341424 entries, 0 to 341423
Data columns (total 24 columns):
reservation_id                        341424 non-null object
booking_date                          341424 non-null object
checkin_date                          341424 non-null object
checkout_date                         341424 non-null object
channel_code                          341424 non-null int64
main_product_code                     341424 non-null int64
numberofadults                        341424 non-null int64
numberofchildren                      341424 non-null int64
persontravellingid                    341424 non-null int64
resort_region_code                    341424 non-null int64
resort_type_code                      341424 non-null int64
room_type_booked_code                 341424 non-null int64
roomnights                            341424 non-null int64
season_holidayed_code                 341310 non-null float64
state_code_residence                  33666

In [6]:
df_test = pd.read_csv("test.csv",parse_dates=True,date_parser=dateparse)
df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 146765 entries, 0 to 146764
Data columns (total 23 columns):
reservation_id              146765 non-null object
booking_date                146765 non-null object
checkin_date                146765 non-null object
checkout_date               146765 non-null object
channel_code                146765 non-null int64
main_product_code           146765 non-null int64
numberofadults              146765 non-null int64
numberofchildren            146765 non-null int64
persontravellingid          146765 non-null int64
resort_region_code          146765 non-null int64
resort_type_code            146765 non-null int64
room_type_booked_code       146765 non-null int64
roomnights                  146765 non-null int64
season_holidayed_code       146730 non-null float64
state_code_residence        144505 non-null float64
state_code_resort           146765 non-null int64
total_pax                   146765 non-null int64
member_age_buckets          146

In [7]:
from sklearn.utils import shuffle
df_train = shuffle(df_train)
df = df_train.append(df_test, ignore_index=1)

/usr/local/lib/python3.6/dist-packages/pandas/core/frame.py:6692: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  sort=sort)


In [0]:
import matplotlib
%matplotlib inline
import seaborn as sns





In [9]:
df.describe()

amount_spent_per_room_night_scaled  booking_type_code   channel_code  \
count                       341424.000000      488189.000000  488189.000000   
mean                             7.720801           1.123952       1.902587   
std                              1.090109           0.329527       0.915112   
min                              1.600397           1.000000       1.000000   
25%                              7.187260           1.000000       1.000000   
50%                              7.860814           1.000000       2.000000   
75%                              8.419144           1.000000       3.000000   
max                             10.816652           2.000000       3.000000   

       main_product_code  numberofadults  numberofchildren  \
count      488189.000000   488189.000000     488189.000000   
mean            2.207459        3.275522          0.362573   
std             0.821989        1.764458          0.758078   
min             1.000000        0.000000          0.000000   
25%             2.000000        2.000000          0.000000   
50%             2.000000        3.000000          0.000000   
75%             3.000000        4.000000          0.000000   
max             7.000000       32.000000         13.000000   

       persontravellingid  resort_region_code  resort_type_code  \
count       488189.000000       488189.000000     488189.000000   
mean           209.955489            1.773303          2.194687   
std            864.891408            0.751550          1.549417   
min             45.000000            1.000000          0.000000   
25%             45.000000            1.000000          1.000000   
50%             45.000000            2.000000          2.000000   
75%             45.000000            2.000000          3.000000   
max           4995.000000            3.000000          7.000000   

       room_type_booked_code     roomnights  season_holidayed_code  \
count          488189.000000  488189.000000          488040.000000   
mean                2.889850       3.735949               2.676035   
std                 0.804194       2.480057               0.884669   
min                 1.000000     -45.000000               1.000000   
25%                 2.000000       2.000000               2.000000   
50%                 3.000000       3.000000               3.000000   
75%                 3.000000       4.000000               3.000000   
max                 6.000000      80.000000               4.000000   

       state_code_residence  state_code_resort      total_pax  
count         481165.000000      488189.000000  488189.000000  
mean               7.258755           4.759472       3.191893  
std                4.717773           3.103048       1.166638  
min                1.000000           1.000000       0.000000  
25%                4.000000           2.000000       2.000000  
50%                8.000000           5.000000       3.000000  
75%                8.000000           7.000000       4.000000  
max               38.000000          13.000000      24.000000

In [10]:
df.head()


amount_spent_per_room_night_scaled booking_date  booking_type_code  \
0                            7.416598     13/05/18                  1   
1                            7.313122     11/06/16                  1   
2                            8.312950     02/09/17                  1   
3                            8.064067     12/04/16                  1   
4                            9.181961     02/12/16                  1   

   channel_code checkin_date checkout_date cluster_code  main_product_code  \
0             2     19/06/18      21/06/18            D                  2   
1             3     25/09/16      30/09/16            E                  2   
2             2     25/12/17      26/12/17            B                  2   
3             1     19/07/16      22/07/16            F                  4   
4             1     18/03/17      20/03/17            F                  1   

  member_age_buckets                                           memberid  ...  \
0                  F  3df0253a624a2947ead6a28ef7e0914a492213ed35ca65...  ...   
1                  H  22595b0b80248eb6d8d0e366761ba139a1cc72c3116aa8...  ...   
2                  B  b828358a65ab24e02bd34f1c62600311ffb6c8e45115fd...  ...   
3                  E  8841d6ba17df8ecbf9aedd2394b8673ecdc3ea318abc4f...  ...   
4                  E  04c2c5678714c951a3ba79d333dfe1577d657a9285b522...  ...   

   reservationstatusid_code  \
0                         A   
1                         A   
2                         A   
3                         C   
4                         A   

                                           resort_id  resort_region_code  \
0  98a3ab7c340e8a033e7b37b6ef9428751581760af67bba...                   2   
1  81b8a03f97e8787c53fe1a86bda042b6f0de9b0ec9c093...                   1   
2  0b918943df0962bc7a1824c0555a389347b4febdc7cf9d...                   2   
3  b17ef6d19c7a5b1ee83b907c595526dcb1eb06db8227d6...                   1   
4  49d180ecf56132819571bf39d9b7b342522a2ac6d23c14...                   1   

  resort_type_code room_type_booked_code roomnights  season_holidayed_code  \
0                2                     4          2                    3.0   
1                5                     3         10                    4.0   
2                4                     4          2                    2.0   
3                1                     3          3                    4.0   
4                1                     2          6                    3.0   

   state_code_residence  state_code_resort  total_pax  
0                   8.0                  2          2  
1                   1.0                  1          2  
2                   8.0                 10          3  
3                   8.0                  6          4  
4                   7.0                  6          5  

[5 rows x 24 columns]

In [11]:
  df["channel_code"].unique()

array([2, 3, 1])

In [12]:

print(df["reservationstatusid_code"].unique())
print(df["persontravellingid"].unique())
print(df["state_code_resort"].unique())
print(df["member_age_buckets"].unique())
print(df["booking_type_code"].unique())
print(df["channel_code"].unique())
print(df["main_product_code"].unique())

['A' 'C' 'B' 'D']
[  45   46   47 4752 4753 4995]
[ 2  1 10  6  5  7  3  9 11  4 13]
['F' 'H' 'B' 'E' 'D' 'G' 'C' 'J' 'I' 'A']
[1 2]
[2 3 1]
[2 4 1 3 7]


cluster code = 6 unique   ['F', 'E', 'D', 'B', 'A', 'C']

```
# This is formatted as code
```
reservationstatusid_code = 4 unique values ['C', 'A', 'B', 'D']

booking_type_code = 2 unique values [1, 2]

member_age_buckets	= 10 unique values ['F', 'H', 'D', 'E', 'I', 'G', 'C', 'B', 'A', 'J']

state_code_resort = 11 unique  values [ 3,  5,  1,  2,  6,  9, 10,  4, 11,  7, 13]

persontravellingid = 6 unique values [  46,   47,   45, 4752, 4753, 4995]

main_product_code = 5 unique values [1, 2, 3, 4, 7]

channel_code = 3 unique values [3, 1, 2]

Find/Derive -----

Total number of days the person has stayed

The month in which booking is made

Check if the booking and check in date is same

Check the difference between the check out and booking date


In [13]:
pd.to_datetime(df["booking_date"][0]).month
from datetime import datetime
d = datetime.strptime(df["checkin_date"][1], '%d/%m/%y')
print((d.date()))

2016-09-25


In [0]:
df.loc[:,"new_checkin"] = df["checkin_date"].apply(lambda x :datetime.strptime(x, '%d/%m/%y').date().strftime('%Y-%m-%d'))
df.loc[:,"new_checkout"] = df["checkout_date"].apply(lambda x :datetime.strptime(x, '%d/%m/%y').date().strftime('%Y-%m-%d'))
df.loc[:,"new_booking"] = df["booking_date"].apply(lambda x :datetime.strptime(x, '%d/%m/%y').date().strftime('%Y-%m-%d'))

In [0]:
df.loc[:,"booking_month"] = df["new_booking"].apply(lambda x: pd.to_datetime(x,format="%Y-%m-%d").month)
df.head(20)

In [0]:
df.loc[:,"checkin_month"] = df["new_checkin"].apply(lambda x : pd.to_datetime(x,format="%Y-%m-%d").month)

In [0]:
df.loc[:,"checkin_bookin"] = pd.to_datetime(df["new_checkin"],format="%Y-%m-%d") - pd.to_datetime(df["new_booking"],format="%Y-%m-%d")
df["checkin_bookin"] = df["checkin_bookin"].dt.days

In [0]:

df.loc[:,"same_day"] = df["checkin_bookin"].apply(lambda x : 1 if x != 0 else 0)

In [0]:
df.loc[:,"days_staying"] = pd.to_datetime(df["new_checkout"],format="%Y-%m-%d") - pd.to_datetime(df["new_checkin"],format="%Y-%m-%d")
df["days_staying"] = df["days_staying"].dt.days

In [0]:
df.head()

In [0]:
redundant_cols = ["booking_date","checkin_date","checkout_date","new_booking","new_checkin","new_checkout"] 

In [0]:
df.drop(redundant_cols,inplace=True,axis=1)
df.to_csv("date_pre.csv",index=False)

In [0]:
df.info()
r_id = df["reservation_id"]
df.drop(["reservation_id","memberid"],inplace=True,axis=1)

In [0]:
df.drop("resort_id",inplace=True,axis=1)
df.info()

In [0]:
df["season_holidayed_code"].fillna(df["season_holidayed_code"].mean(),inplace=True)


In [0]:
df["state_code_residence"].fillna(df["state_code_residence"].mean(),inplace=True)
df["persontravellingid"] = df["persontravellingid"].astype("object",inplace=True)

In [0]:
df_dumm = pd.get_dummies(df)
columns = list(df.columns)
df_dumm.info()

In [0]:
df_dumm.head()

In [0]:
y = df_dumm["amount_spent_per_room_night_scaled"].values
X = df_dumm.drop("amount_spent_per_room_night_scaled",axis=1).values
X_train, X_test  =  X[:len(df_train)], X[len(df_train):]
y_train, y_test  =  y[:len(df_train)], y[len(df_train):]

In [0]:
for i in columns:
  sns.distplot(df["total_pax"],kde=False)

In [0]:
import numpy as np
def out_detect(data):
	out_dict = {}
	for i in data.describe().columns:
		ltv=0
		utv=0
		q1 = data.describe().at['25%',i]
		q3 = data.describe().at['75%',i]
		iqr = q3-q1
		ltv = q1-1.5*iqr
		utv = q3 +1.5*iqr
		print(q1,q3)
		print(ltv,utv)
		x = np.array(data[i])
		p=[]
		count = 0 
		for j in x:
				if j>utv or j<ltv:
					count= count+1
					p.append(round(data[i].median(),2))

				else:
					p.append(j)
		out_dict[i] = str(count)
		data[i] = p 
		print("for columns {} count is {}".format(i,count))
	print(out_dict)
	return(data)

In [0]:
def corr_remove(data):
	col_corr=set()
	corr_matrix = data.corr()
	for i in range(len(corr_matrix.columns)):
		for j in range(i):
			if abs(corr_matrix.iloc[i, j])>threshold:
				colname = corr_matrix.columns[i]
				col_corr.add(colname)
	return(col_corr)

In [0]:
!pip install catboost -U

In [0]:
y_test

In [0]:
# from catboost import CatBoostRegressor, CatBoostClassifier


# classifier = CatBoostRegressor(iterations = 10000,
#                                 eval_metric='RMSE',
#                                 loss_function= 'RMSE',
#                                 cat_features=[3, 4, 5, 6, 8, 9, 10, 11, 12, 13, 14,20,24,28,30,34,37,39],
#                                 #learning_rate=0.003,
#                                 #depth=7
#                                 )

# classifier.fit(X_train, y_train)

In [0]:
from sklearn.model_selection import cross_val_predict, GridSearchCV
from sklearn.ensemble import RandomForestRegressor
from sklearn.preprocessing import MinMaxScaler

gsc = GridSearchCV(
        estimator=RandomForestRegressor(),
        param_grid={
            'max_depth': (3,7),
            'n_estimators': (80 ,100),
        },
        cv=5, scoring='neg_mean_squared_error', verbose=30,n_jobs=-1)

In [0]:
# y = df_dumm["amount_spent_per_room_night_scaled"]
# X_df = df_dumm.drop("amount_spent_per_room_night_scaled",axis=1)
# X_df.columns

In [0]:
grid_result = gsc.fit(X_train, y_train)
best_params = grid_result.best_params_

In [0]:
from sklearn.model_selection import cross_val_score
rfr = RandomForestRegressor(max_depth=best_params["max_depth"], n_estimators=best_params["n_estimators"],random_state=False, verbose=10)

scores = cross_val_score(rfr, X_train, y_train, cv=10, scoring='neg_mean_absolute_error')
print(scores)

In [0]:
from sklearn.externals import joblib
rfr = RandomForestRegressor(max_depth=best_params["max_depth"], n_estimators=best_params["n_estimators"],random_state=101, verbose=10)
rfr.fit(X_train,y_train)
joblib.dump(rfr,"rfr.pkl")

In [0]:
from google.colab import files
files.download('rfr.pkl') 

In [0]:
from sklearn.externals import joblib
rfr = joblib.load("rfr.pkl")
rfr_predict = rfr.predict(X_test)

In [0]:
final = pd.DataFrame()
r_id = df_test["reservation_id"]
final.loc[:,"reservation_id"]=r_id
final.loc[:,"amount_spent_per_room_night_scaled"]=rfr_predict

In [0]:
final.to_csv("result.csv",index=False)

In [0]:
files.download("result.csv")